tee-worker/omni-executor/intent/executors/cross-chain/src/lib.rs

```diff
@@ -15,31 +15,176 @@
+		RpcClient: SubstrateRpcClient<Header>,
+		RpcClientFactory: SubstrateRpcClientFactory<Header, RpcClient>,
+	> CrossChainIntentExecutor<Header, RpcClient, RpcClientFactory>
+{
+	pub fn new(
+		parentchain_rpc_client_factory: Arc<RpcClientFactory>,
+		transaction_signer: Arc<ParentchainTxSigner>,
+		rpc_endpoint_registry: RpcEndpointRegistry,
+		pumpx_signer_client: Arc<SignerClient>,
+	) -> Result<Self, ()> {
+		let account_asset_lock = AccountAssetLocks::<AccountWideAssetsLock>::empty();
+		Ok(Self {
+			parentchain_rpc_client_factory,
+			transaction_signer,
+			account_asset_lock,
+			rpc_endpoint_registry,
+			pumpx_signer_client,
+			phantom: PhantomData,
+		})
 	}
 }
 
 #[async_trait]
-impl IntentExecutor for CrossChainIntentExecutor {
-	async fn execute(&self, _account_id: &AccountId, intent: Intent) -> Result<(), ()> {
+impl<
+		Header: Send + Sync,
+		RpcClient: SubstrateRpcClient<Header> + Send + Sync,
+		RpcClientFactory: SubstrateRpcClientFactory<Header, RpcClient> + Send + Sync,
+	> IntentExecutor for CrossChainIntentExecutor<Header, RpcClient, RpcClientFactory>
+{
+	async fn execute(
+		&self,
+		account_id: &AccountId,
+		intent_id: IntentId,
+		intent: Intent,
+	) -> Result<(), ()> {
 		match intent {
-			Intent::Swap(_swap_order, _ccsp, _scsp) => {
+			Intent::Swap(ref swap_order, ref _ccsp, ref _scsp) => {
+				let Ok(mut rpc_client) = self.parentchain_rpc_client_factory.new_client().await
+				else {
+					log::error!("Failed to create rpc client");
+					return Err(());
+				};
+				let available_amount = match &swap_order.from_asset {
+					ChainAsset::Ethereum(chain_id, token) => {
+						let rpc_url =
+							self.rpc_endpoint_registry.get(&Chain::Ethereum(*chain_id)).ok_or(())?;
+						let address = self
+							.pumpx_signer_client
+							.request_wallet(
+								pumpx::signer_client::ChainType::Evm,
+								0,
+								*account_id.as_ref(),
+							)
+							.await
+							.map_err(|e| {
+								error!("Could not get wallet from pumpx-signer: {:?}", e)
+							})?;
+						query_ethereum(rpc_url, EthereumAddress::from_slice(&address), token)
+							.await?
+					},
+					ChainAsset::Solana(token) => {
+						let rpc_url = self.rpc_endpoint_registry.get(&Chain::Solana).ok_or(())?;
+						let address = self
+							.pumpx_signer_client
+							.request_wallet(
+								pumpx::signer_client::ChainType::Solana,
+								0,
+								*account_id.as_ref(),
+							)
+							.await
+							.map_err(|e| {
+								error!("Could not get wallet from pumpx-signer: {:?}", e)
+							})?;
+						let pubkey = SolanaPubkey::try_from(address).map_err(|e| {
+							error!("Could not create solana pubkey from wallet address: {:?}", e)
+						})?;
+						query_solana(rpc_url, &pubkey, token).await.map(|v| AmountType::from(v))?
+					},
+				};
+				self.account_asset_lock.check_and_insert(
+					account_id.clone(),
+					swap_order.from_asset.clone(),
+					AmountType::from(swap_order.from_amount),
+					available_amount,
+				)?;
```

Comment:
I think we agreed that we don't have to lock the asset (as we do ordered processing: 1. deduct balance 2. binance convert 3. payout), but just need to put the next cross-chain swap request on-hold until this payout is done.

Or did you plan to do it in another PR?

Reply:
I'm also thinking of a corner case:

1. user requests to do a cross-chain swap with intent-id 1
2. user requests to do a single-chain swap with intent-id 2
3. user requests to do a single-chains wap with intent-id 3
4. user requests to do a cross-chains wap with intent-id 4

In this case, shall we block 2 and 3 until 1 is finished? Probably not.

But then it means:
- we should only lock the cross-chain trade processing part (so we'll need to distinguish between cross-chain swap and single-chain swap)
- the id continuity check isn't simple +1 check, we need to keep track of last unfinished cross-chain trade id (e.g. for intent 4 we need check if payout of intent 1 has finished)

Is it a bit too complicated? 😂

Reply:
> I think we agreed that we don't have to lock the asset (as we do ordered processing: 1. deduct balance 2. binance convert 3. payout), but just need to put the next cross-chain swap request on-hold until this payout is done.
> 
> Or did you plan to do it in another PR?

Currently we are using `AccountWideAssetsLock` which simply locks whole account for the time of intent processing (only one intent at the time can be processed per account). 

Worker:
1. checks intent id
2. updates intent id
3. checks account lock

Reply:
I misread it sorry, I see it's an account-wide lock. 

But it also means if the next request comes, it will error out instead of waiting, right? Acutally I'm rethinking if we even need a lock here. Can we simply continue processing, if the balance is insufficient, it will error out anyway when we try to deduct user balance?

Reply:
We can queue intents -> persist them to disk.
Upon intent processing finish - we can check if there is any pending intent and send if for processing.
Some intents can be queue - some not - depends on type <-but this adds another level of complexity. Once intent type is allowed for parallel processing you need to keep it in mind everytime you make any changes to that type processing logic, otherwise you can bring troubles 😈